### PR TITLE
Feature/text offscreen canvas measure

### LINKED ISSUE #1364: Text formatting like `rightAlign` is only available when `fixedWidth` is used

### PR DESCRIPTION
**Pull request by merryman**
Implements #1379.
Resolves #1364.

### CODE REVIEW

**Comment by linusha** [comment]
Still in the workings: a small bug with the rendering of the cursor/selections in wrapped lines.

File: lively.morphic/layout.js [hunk 1]
```diff
@@ -1263,7 +1294,7 @@ export class TilingLayout extends Layout {
         margin.top = 0;
       }
     } else {
-      yogaNode.setHeight(submorph.height); // what if the text morph layout is fitting the morph here?
+      yogaNode.setHeight(Math.floor(submorph.height)); // what if the text morph layout is fitting the morph here?
```
Comment (linusha) [inline lively.morphic/layout.js:1297]: Is that comment still relevant? From what I understand it is not?

File: lively.morphic/rendering/font-metric.js [hunk 2]
```diff
@@ -329,6 +348,22 @@ class DOMTextMeasure {
     style.overflow = isRoot && !this.debug ? 'hidden' : 'visible';
   }
 
+  canBeMeasuredViaCanvas (aMorph) {
+    if (!aMorph.allFontsLoaded()) return false;
+    const { fontFamily, fontWeight, fontStyle } = aMorph;
+    const key = `${fontFamily}-${fontWeight}-${fontStyle}`;
+    if (key in this.canvasCompatibility) return this.canvasCompatibility[key];
+
+    // determine wether or not a font does some werid shit with letter spacing
```
Comment (linusha) [inline lively.morphic/rendering/font-metric.js:357]:
```suggestion
    // determine wether or not a font does some weird shit with letter spacing
    // note, that this is a heuristic only
```

File: lively.morphic/rendering/font-metric.js [hunk 3]
```diff
@@ -329,6 +348,22 @@ class DOMTextMeasure {
...
+    const errorMargin = 3;
+    // this.env.fontMetric._domMeasure.canvasCompatibility = {}
```
Comment (linusha) [inline lively.morphic/rendering/font-metric.js:359]:
```suggestion
```

File: lively.morphic/layout.js [hunk 4]
```diff
@@ -840,15 +858,27 @@ export class TilingLayout extends Layout {
...
+
+    if (!isPreliminary && widthPolicy === 'fill' && String(newWidth) !== 'NaN' && newWidth !== morph.width) {
+      const widthBefore = morph.width;
+      morph.withMetaDo({ isLayoutAction: true, skipRender: false }, () => morph.width = newWidth);
+      if (morph.isText && !morph.fixedHeight) {
+        retriggerLayout = true;
+      }
+      if (morph.isText && !morph.fixedWidth) {
+        console.warn(`The text morph ${morph.id} is set to hug its contents yet configure to fill the container horizontally!`);
+      }
```
Comment (linusha) [inline lively.morphic/layout.js:871]: This one gets really spammy, even when only playing around for a bit on the new branch...

**Review by linusha — CHANGES_REQUESTED** [review]

File: lively.morphic/layout.js (hunk 4 above)
Comment (merryman) [reply lively.morphic/layout.js:871]: Which goes to show that there are a lot of nonsensical configurations in the system. Would you prefer a default behavior? If so which one should win?

**Review by linusha — APPROVED** [review]

**Pull request merged**